metamodel/processor/src/main/java/org/hibernate/search/metamodel/processor/impl/IndexedEntityMetamodelAnnotationProcessor.java

```diff
@@ -0,0 +1,150 @@
+		this.context = context;
+		this.introspectorContext = new ProcessorIntrospectorContext( context, new HashMap<>() );
+	}
+
+	@Override
+	public void process(RoundEnvironment roundEnv) {
+		TypeElement indexedAnnotation = context.elementUtils().getTypeElement( ANNOTATION_INDEXED );
+		Set<? extends Element> indexedEntities = roundEnv.getElementsAnnotatedWith( indexedAnnotation );
+
+		try ( CloseableSearchMapping searchMapping = StandalonePojoIntegrationBooter.builder()
+				.introspectorCustomizer( this::wrapIntrospector )
+				.property( "hibernate.search.backend.directory.type", "local-heap" )
+				.property(
+						StandalonePojoMapperSettings.MAPPING_CONFIGURER,
+						BeanReference.ofInstance( (StandalonePojoMappingConfigurer) configurationContext -> {
+							// disable any discovery by annotations -- that won't work as we do not implement annotations in the pojo model
+							configurationContext.annotationMapping()
+									.discoverAnnotatedTypesFromRootMappingAnnotations( false )
+									.discoverAnnotationsFromReferencedTypes( false )
+									.discoverJandexIndexesFromAddedTypes( false )
+									.buildMissingDiscoveredJandexIndexes( false );
+
+							ProgrammaticMappingConfigurationContext programmaticMapping =
+									configurationContext.programmaticMapping();
+
+							ProcessorAnnotationProcessorContext ctx = new ProcessorAnnotationProcessorContext(
+									context.elementUtils(),
+									context.typeUtils(), context.messager(), programmaticMapping
+							);
+
+							for ( Element el : indexedEntities ) {
+								if ( el.getKind().isClass() || el.getKind().isInterface() ) {
+									TypeElement indexedEntityType = (TypeElement) el;
+
+									String typeName = indexedEntityType.getQualifiedName().toString();
+
+									introspectorContext.typeElementsByName().clear();
+									introspectorContext.typeElementsByName().put( typeName, indexedEntityType );
+
+									TypeMappingStep typeMappingContext = programmaticMapping.type( typeName );
+									typeMappingContext.indexed().enabled( true );
+									typeMappingContext.searchEntity().name( typeName );
+
+									processTypeAndProperties( indexedEntityType, typeMappingContext, ctx );
+								}
+								else {
+									// TODO: generate message bundle with JBoss Logging ?
+									context.messager()
+											.printMessage(
+													Diagnostic.Kind.ERROR,
+													"Unexpected location of the " + indexedAnnotation.getQualifiedName()
+															+ ". Expected to be placed on an indexed type.",
+													el
+											);
+								}
+							}
+
+						} )
+				).build()
+				.boot() ) {
+			for ( SearchIndexedEntity<?> entity : searchMapping.allIndexedEntities() ) {
+				context.messager().printMessage( Diagnostic.Kind.NOTE, entity.name() );
+				entity.indexManager().descriptor().staticFields()
+						.forEach( f -> context.messager().printMessage( Diagnostic.Kind.NOTE, f.toString() ) );
+			}
```

Comment:
for now I just print the fields out so it's easy to see what is produced, but I've already started looking into how to generate classes. From what I see, Hibernate ORM/Tools are writing the classes with custom code, I also found this https://github.com/palantir/javapoet and was considering whether it would make sense to pull a dependency, or just do yet another custom class writer 😄

Reply:
If it's not too much work, yes I'd start with custom code, possibly copy-pasted. It doesn't look very complex nor likely to require much maintenance, considering our needs?